pkg/virtctl/adm/logverbosity/logverbosity.go

```diff
@@ -0,0 +1,447 @@
+	// set default verbosity first
+	// it is used to fill the unattended verbosity with default verbosity
+	// key: JSONName, value: verbosity
+	var verbosityVal = map[string]uint{
+		"virtAPI":        virtconfig.DefaultVirtAPILogVerbosity,
+		"virtController": virtconfig.DefaultVirtControllerLogVerbosity,
+		"virtHandler":    virtconfig.DefaultVirtHandlerLogVerbosity,
+		"virtLauncher":   virtconfig.DefaultVirtLauncherLogVerbosity,
+		"virtOperator":   virtconfig.DefaultVirtOperatorLogVerbosity,
+	}
+
+	// if verbosity has been set in the KubeVirt CR, use the verbosity
+	lvMap, err := hasVerbosityInKV(kv)
+	if err != nil {
+		return nil, err
+	}
+	if len(lvMap) > 0 {
+		for key, value := range lvMap {
+			verbosityVal[key] = value
+		}
+	}
+
+	// create a message to show verbosity for the specified component
+	messages := createOutputMessage(verbosityVal, options)
+
+	return messages, nil
+}
+
+func setVerbosity(lvMap map[string]uint, options map[string]uint, patchData *[]patch.PatchOperation, op *string, path *string) {
+	// update lvMap based on the user-specified verbosity for all components
+	if verbosity, exist := options["all"]; exist {
+		for component := virtAPI; component < all; component++ {
+			JSONName := getJSONNameByVirtComponent(component)
+			lvMap[JSONName] = verbosity
+		}
+	}
+	// update lvMap based on the user-specified verbosity for each component
+	for componentName, verbosity := range options {
+		if componentName == "all" {
+			continue
+		}
+		JSONName := getJSONNameByComponentName(componentName)
+		lvMap[JSONName] = verbosity
+	}
+
+	if len(lvMap) != 0 {
+		addPatch(patchData, op, path, lvMap)
+	}
+}
+
+func addPatch(patchData *[]patch.PatchOperation, op *string, path *string, lvMap map[string]uint) {
+	*patchData = append(*patchData, patch.PatchOperation{
+		Op:    *op,
+		Path:  *path,
+		Value: lvMap,
+	})
+}
+
+func resetVerbosity(lvMap map[string]uint, patchData *[]patch.PatchOperation, op *string, path *string) {
+	// reset only if verbosity exists, otherwise do nothing
+	if len(lvMap) != 0 {
+		// add an empty object (removing the logVerbosity field can be another method)
+		emptyMap := map[string]uint{} // does not change the caller's lvMap
+		addPatch(patchData, op, path, emptyMap)
+	}
+}
+
+func createPatch(kv *v1.KubeVirt, options map[string]uint) ([]byte, error) {
+	patchData := []patch.PatchOperation{}
+	// just "add" is fine, no need of "replace" and "remove"
+	// https://www.rfc-editor.org/rfc/rfc6902
+	op := patch.PatchAddOp
+	path := "/spec/configuration/developerConfiguration/logVerbosity"
+
+	// if there is a logVerbosity field in the KubeVirt CR, fill in the data in the lvMap
+	lvMap, err := hasVerbosityInKV(kv)
+	if err != nil {
+		return nil, err
+	}
+	if lvMap == nil {
+		// if map is nil (logVerbosity field in the KubeVert CR is nil), need initialization
+		lvMap = make(map[string]uint)
+	}
+
+	if isReset {
+		resetVerbosity(lvMap, &patchData, &op, &path)
+		lvMap = map[string]uint{}
+	}
```

Comment:
Note that you can potentially create lvMap twice. Just move `if lvMap == nil` after `if isReset`

Reply:
Agreed.

The `lvMap` argument of the `resetVerbosity` function was set to **pass by value** instead of **pass by reference** even though it is a map, so the `lvMap` could not be updated within `resetVerbosity` function. (The reason why it was set to pass by value is unknown.)

You can make it a pointer (`lvMap *map[string]uint`) and pass by reference, but then `golangci-lint` will point it out.
```
pkg/virtctl/adm/logverbosity/logverbosity.go:305:21: ptrToRefParam: consider `currentLv' to be of non-pointer type (gocritic)
func resetVerbosity(currentLv *map[string]uint, patchData *[]patch.PatchOperation) {
                    ^
```
Then, I put `lvMap = map[string]unit{}` here.
```
325	if isReset {
		resetVerbosity(lvMap, &patchData, &op, &path)
		lvMap = map[string]uint{}
	}
```

However, this code itself is not necessarily enough to make it a function, so we will stop assuming that `resetVerbosity` is a function.


On the other hand, `lvMap = make(map[string]unit)` below has nothing to do with the logic of `reset` operation.
```
320	if lvMap == nil {
		// if map is nil (logVerbosity field in the KubeVert CR is nil), need initialization
		lvMap = make(map[string]uint)
	}
```
It is initialized to avoid an error if `lvMap=nil` when accessing the `lvMap` key in the `setVerbosity` function, in the following for example.
```
271	if verbosity, exist := options["all"]; exist {
		for component := virtAPI; component < all; component++ {
			JSONName := getJSONNameByVirtComponent(component)
			lvMap[JSONName] = verbosity
		}
	}
```
Therefore, I think it is appropriate to place this code (`lvMap = make(map[string]unit)`)  immediately after `currentLv, err := hasVerbosityInKV(kv)` or just before `setVerbosity(currentLv, options, &patchData)`, which actually relates to the code.

For now, I'll move `currentLv = make(map[string]uint)` to `main` with `currentLv, err := hasVerbosityInKV(kv)` like below, and push the code.
```
	currentLv, err := hasVerbosityInKV(kv)
	if err != nil {
		return err
	}
	// if map is nil (= logVerbosity field in the KubeVert CR is nil), need initialization to access a key in the map
	if currentLv == nil {
		currentLv = make(map[string]uint)
	}
	// create patch data
	patchData, err := createPatch(currentLv, options)
	if err != nil {
		return err
	}
```

If you think this is not a good idea, please let me know.

Reply:
> The lvMap argument of the resetVerbosity function was set to pass by value instead of pass by reference even though it is a map, so the lvMap could not be updated within resetVerbosity function.

Map types are similar to channels, that is:

```Go
func add(m map[string]bool) {
    m["foo"] = true
}

func main() {
     m := make(map[string]bool)
     add(m)
     fmt.Printf("%v\n", m)
}
```

Prints `map[foo:true]`.

The map needs to be created though.

> If you think this is not a good idea, please let me know.

I think it is fine :)

Reply:
Thank you for the explanation.

Reply:
> However, this did not work. When I changed the lvMap (lvMap = map[string]uint{}) in the resetVerbosity function

You created a new map in the inner function and the caller does not know anything about it.

```Go
func create(m map[string]string) {
    fmt.Printf("create: %v\n", m)
    m = map[string]string{"drink": "wine"}
    fmt.Printf("create: %v\n", m)
}

func main() {
    m := map[string]string{"food": "pizza"}
    fmt.Printf("main  : %v\n", m)
    create(m)
    fmt.Printf("main  : %v\n", m)
}
```

```
main  : map[food:pizza]
create: map[food:pizza]
create: map[drink:wine]
main  : map[food:pizza]
```

The variable `m` in `main()` and the variable `m` in `create()`, they have different addresses but that does not matter as they `point` to the same memory region where the `map` was created in the `main()` function. That changed when we create a new map in the `create()`  function, as `m`  from `create()`  is now pointing to a different address, a different map.

Reply:
Makes sense. Thanks a lot!